In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/nonameexi/screening-dataset/Scrrening.csv


In [2]:
import pandas as pd
df = pd.read_csv("/kaggle/input/datasets/nonameexi/screening-dataset/Scrrening.csv")

In [3]:
df

,job_id,title,location,department,salary_range,company_profile,description,requirements,benefits,telecommuting,has_company_logo,has_questions,employment_type,required_experience,required_education,industry,function,fraudulent
0,1,Marketing Intern,"US, NY, New York",Marketing,NaN,"We're Food52, and we've created a groundbreaki...","Food52, a fast-growing, James Beard Award-winn...",Experience with content management systems a m...,NaN,0,1,0,Other,Internship,NaN,NaN,Marketing,0
1,2,Customer Service - Cloud Video Production,"NZ, , Auckland",Success,NaN,"90 Seconds, the worlds Cloud Video Production ...",Organised - Focused - Vibrant - Awesome!Do you...,What we expect from you:Your key responsibilit...,What you will get from usThrough being part of...,0,1,0,Full-time,Not Applicable,NaN,Marketing and Advertising,Customer Service,0
2,3,Commissioning Machinery Assistant (CMA),"US, IA, Wever",NaN,NaN,Valor Services provides Workforce Solutions th...,"Our client, located in Houston, is actively se...",Implement pre-commissioning and commissioning ...,NaN,0,1,0,NaN,NaN,NaN,NaN,NaN,0
3,4,Account Executive - Washington DC,"US, DC, Washington",Sales,NaN,Our passion for improving quality of life thro...,THE COMPANY: ESRI – Environmental Systems Rese...,"EDUCATION: Bachelor’s or Master’s in GIS, busi...",Our culture is anything but corporate—we have ...,0,1,0,Full-time,Mid-Senior level,Bachelor's Degree,Computer Software,Sales,0
4,5,Bill Review Manager,"US, FL, Fort Worth",NaN,NaN,SpotSource Solutions LLC is a Global Human Cap...,JOB TITLE: Itemization Review ManagerLOCATION:...,QUALIFICATIONS:RN license in the State of Texa...,Full Benefits Offered,0,1,1,Full-time,Mid-Senior level,Bachelor's Degree,Hospital & Health Care,Health Care Provider,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
17875,17876,Account Director - Distribution,"CA, ON, Toronto",Sales,NaN,Vend is looking for some awesome new talent to...,Just in case this is the first time you’ve vis...,To ace this role you:Will eat comprehensive St...,What can you expect from us?We have an open cu...,0,1,1,Full-time,Mid-Senior level,NaN,Computer Software,Sales,0
17876,17877,Payroll Accountant,"US, PA, Philadelphia",Accounting,NaN,WebLinc is the e-commerce platform and service...,The Payroll Accountant will focus primarily on...,- B.A. or B.S. in Accounting- Desire to have f...,Health &amp; WellnessMedical planPrescription ...,0,1,1,Full-time,Mid-Senior level,Bachelor's Degree,Internet,Accounting/Auditing,0
17877,17878,Project Cost Control Staff Engineer - Cost Con...,"US, TX, Houston",NaN,NaN,We Provide Full Time Permanent Positions for m...,Experienced Project Cost Control Staff Enginee...,At least 12 years professional experience.Abil...,NaN,0,0,0,Full-time,NaN,NaN,NaN,NaN,0
17878,17879,Graphic Designer,"NG, LA, Lagos",NaN,NaN,NaN,Nemsia Studios is looking for an experienced v...,1. Must be fluent in the latest versions of Co...,Competitive salary (compensation will be based...,0,0,1,Contract,Not Applicable,Professional,Graphic Design,Design,0


In [4]:
print("Null Values")
df.isna().sum()

Null Values


job_id                     0
title                      0
location                 346
department             11547
salary_range           15012
company_profile         3308
description                1
requirements            2696
benefits                7212
telecommuting              0
has_company_logo           0
has_questions              0
employment_type         3471
required_experience     7050
required_education      8105
industry                4903
function                6455
fraudulent                 0
dtype: int64

In [5]:
df.duplicated().sum()

np.int64(0)

In [6]:
df.shape

(17880, 18)

In [7]:
df.fillna("Unknown", inplace=True)

In [8]:
df.isna().sum()

job_id                 0
title                  0
location               0
department             0
salary_range           0
company_profile        0
description            0
requirements           0
benefits               0
telecommuting          0
has_company_logo       0
has_questions          0
employment_type        0
required_experience    0
required_education     0
industry               0
function               0
fraudulent             0
dtype: int64

In [9]:
print("Shape:", df.shape)

print("\nDuplicate rows:", df.duplicated().sum())

print("\nTarget distribution:")
print(df["fraudulent"].value_counts())

print("\nTarget percentage:")
print(df["fraudulent"].value_counts(normalize=True) * 100)

Shape: (17880, 18)

Duplicate rows: 0

Target distribution:
fraudulent
0    17014
1      866
Name: count, dtype: int64

Target percentage:
fraudulent
0    95.1566
1     4.8434
Name: proportion, dtype: float64


# Target is Highly IMbalanced 
# 95% real Jobs
# 5% Fraud

In [10]:
categorical_cols = [
    "employment_type",
    "required_experience",
    "required_education",
    "industry",
    "function",
    "telecommuting",
    "has_company_logo",
    "has_questions"
]

for col in categorical_cols:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False).head(15))


--- employment_type ---
employment_type
Full-time    11620
Unknown       3471
Contract      1524
Part-time      797
Temporary      241
Other          227
Name: count, dtype: int64

--- required_experience ---
required_experience
Unknown             7050
Mid-Senior level    3809
Entry level         2697
Associate           2297
Not Applicable      1116
Director             389
Internship           381
Executive            141
Name: count, dtype: int64

--- required_education ---
required_education
Unknown                              8105
Bachelor's Degree                    5145
High School or equivalent            2080
Unspecified                          1397
Master's Degree                       416
Associate Degree                      274
Certification                         170
Some College Coursework Completed     102
Professional                           74
Vocational                             49
Some High School Coursework            27
Doctorate                          

In [11]:
text_cols = [
    "title",
    "company_profile",
    "description",
    "requirements",
    "benefits"
]

for col in text_cols:
    print(f"\n--- {col} ---")
    print("Missing:", df[col].isna().sum())
    print("Unique:", df[col].nunique())
    print("Average characters:", df[col].fillna("").str.len().mean())


--- title ---
Missing: 0
Unique: 11231
Average characters: 28.530313199105144

--- company_profile ---
Missing: 0
Unique: 1710
Average characters: 622.1965324384787

--- description ---
Missing: 0
Unique: 14802
Average characters: 1218.0048098434004

--- requirements ---
Missing: 0
Unique: 11968
Average characters: 591.1876957494408

--- benefits ---
Missing: 0
Unique: 6205
Average characters: 211.71918344519017


In [12]:
for col in ["telecommuting", "has_company_logo", "has_questions"]:
    print(f"\n--- {col} vs fraudulent ---")
    print(pd.crosstab(
        df[col],
        df["fraudulent"],
        normalize="index"
    ) * 100)


--- telecommuting vs fraudulent ---
fraudulent             0         1
telecommuting                     
0              95.313504  4.686496
1              91.655802  8.344198

--- has_company_logo vs fraudulent ---
fraudulent                0          1
has_company_logo                      
0                 84.071038  15.928962
1                 98.009845   1.990155

--- has_questions vs fraudulent ---
fraudulent             0         1
has_questions                     
0              93.221831  6.778169
1              97.156506  2.843494


In [13]:
cat_cols = [
    "employment_type",
    "required_experience",
    "required_education",
    "industry",
    "function"
]

for col in cat_cols:
    print(f"\n--- {col} vs fraudulent ---")
    result = pd.crosstab(
        df[col],
        df["fraudulent"],
        normalize="index"
    ) * 100

    print(result.sort_values(1, ascending=False).head(15))


--- employment_type vs fraudulent ---
fraudulent               0         1
employment_type                     
Part-time        90.715182  9.284818
Unknown          93.056756  6.943244
Other            93.392070  6.607930
Full-time        95.783133  4.216867
Contract         97.112861  2.887139
Temporary        99.170124  0.829876

--- required_experience vs fraudulent ---
fraudulent                   0         1
required_experience                     
Executive            92.907801  7.092199
Entry level          93.362996  6.637004
Unknown              93.829787  6.170213
Not Applicable       94.623656  5.376344
Director             95.629820  4.370180
Mid-Senior level     97.033342  2.966658
Internship           97.375328  2.624672
Associate            98.171528  1.828472

--- required_education vs fraudulent ---
fraudulent                                  0          1
required_education                                      
Some High School Coursework         25.925926  74.074074

In [14]:
cat_cols = [
    "employment_type",
    "required_experience",
    "required_education",
    "industry",
    "function"
]

for col in cat_cols:
    print(f"\n--- {col} vs fraudulent ---")

    result = pd.crosstab(
        df[col],
        df["fraudulent"],
        normalize="index"
    ) * 100

    print(result.sort_values(1, ascending=False).head(15))


--- employment_type vs fraudulent ---
fraudulent               0         1
employment_type                     
Part-time        90.715182  9.284818
Unknown          93.056756  6.943244
Other            93.392070  6.607930
Full-time        95.783133  4.216867
Contract         97.112861  2.887139
Temporary        99.170124  0.829876

--- required_experience vs fraudulent ---
fraudulent                   0         1
required_experience                     
Executive            92.907801  7.092199
Entry level          93.362996  6.637004
Unknown              93.829787  6.170213
Not Applicable       94.623656  5.376344
Director             95.629820  4.370180
Mid-Senior level     97.033342  2.966658
Internship           97.375328  2.624672
Associate            98.171528  1.828472

--- required_education vs fraudulent ---
fraudulent                                  0          1
required_education                                      
Some High School Coursework         25.925926  74.074074

In [15]:
text_cols = [
    "title",
    "company_profile",
    "description",
    "requirements",
    "benefits"
]

df["combined_text"] = df[text_cols].fillna("").agg(" ".join, axis=1)

print(df["combined_text"].head())
print("\nAverage text length:",
      df["combined_text"].str.len().mean())

0    Marketing Intern We're Food52, and we've creat...
1    Customer Service - Cloud Video Production 90 S...
2    Commissioning Machinery Assistant (CMA) Valor ...
3    Account Executive - Washington DC Our passion ...
4    Bill Review Manager SpotSource Solutions LLC i...
Name: combined_text, dtype: object

Average text length: 2675.638534675615


In [16]:
print(df[["title", "combined_text", "fraudulent"]].head(3))

                                       title  \
0                           Marketing Intern   
1  Customer Service - Cloud Video Production   
2    Commissioning Machinery Assistant (CMA)   

                                       combined_text  fraudulent  
0  Marketing Intern We're Food52, and we've creat...           0  
1  Customer Service - Cloud Video Production 90 S...           0  
2  Commissioning Machinery Assistant (CMA) Valor ...           0  


In [17]:
text_cols = [
    "title",
    "company_profile",
    "description",
    "requirements",
    "benefits"
]

df["combined_text"] = df[text_cols].fillna("").agg(" ".join, axis=1)

print(df["combined_text"].head())
print("\nAverage text length:",
      df["combined_text"].str.len().mean())

0    Marketing Intern We're Food52, and we've creat...
1    Customer Service - Cloud Video Production 90 S...
2    Commissioning Machinery Assistant (CMA) Valor ...
3    Account Executive - Washington DC Our passion ...
4    Bill Review Manager SpotSource Solutions LLC i...
Name: combined_text, dtype: object

Average text length: 2675.638534675615


In [18]:
print(df[["title", "combined_text", "fraudulent"]].head(3))


                                       title  \
0                           Marketing Intern   
1  Customer Service - Cloud Video Production   
2    Commissioning Machinery Assistant (CMA)   

                                       combined_text  fraudulent  
0  Marketing Intern We're Food52, and we've creat...           0  
1  Customer Service - Cloud Video Production 90 S...           0  
2  Commissioning Machinery Assistant (CMA) Valor ...           0  


In [19]:
from sklearn.model_selection import train_test_split

X = df["combined_text"]
y = df["fraudulent"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training samples:", len(X_train))
print("Testing samples:", len(X_test))

print("\nTrain distribution:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTest distribution:")
print(y_test.value_counts(normalize=True) * 100)

Training samples: 14304
Testing samples: 3576

Train distribution:
fraudulent
0    95.155201
1     4.844799
Name: proportion, dtype: float64

Test distribution:
fraudulent
0    95.162192
1     4.837808
Name: proportion, dtype: float64


In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(
    lowercase=True,
    stop_words="english",
    ngram_range=(1, 2),
    min_df=2,
    max_df=0.95,
    sublinear_tf=True
)

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print("Training TF-IDF shape:", X_train_tfidf.shape)
print("Testing TF-IDF shape:", X_test_tfidf.shape)

Training TF-IDF shape: (14304, 376290)
Testing TF-IDF shape: (3576, 376290)


In [21]:
from sklearn.svm import LinearSVC

svm = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm.fit(X_train_tfidf, y_train)

y_pred = svm.predict(X_test_tfidf)

In [22]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_auc_score,
    average_precision_score
)

print("Classification Report:")
print(classification_report(y_test, y_pred, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# LinearSVC decision scores
y_score = svm.decision_function(X_test_tfidf)

print("\nROC-AUC:",
      roc_auc_score(y_test, y_score))

print("PR-AUC:",
      average_precision_score(y_test, y_score))

Classification Report:
              precision    recall  f1-score   support

           0     0.9933    0.9994    0.9963      3403
           1     0.9868    0.8671    0.9231       173

    accuracy                         0.9930      3576
   macro avg     0.9901    0.9332    0.9597      3576
weighted avg     0.9930    0.9930    0.9928      3576


Confusion Matrix:
[[3401    2]
 [  23  150]]

ROC-AUC: 0.9955836315797519
PR-AUC: 0.9629316538762619


In [23]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from scipy.sparse import hstack, csr_matrix

categorical_cols = [
    "employment_type",
    "required_experience",
    "required_education",
    "industry",
    "function"
]

binary_cols = [
    "telecommuting",
    "has_company_logo",
    "has_questions"
]

# Use the same train/test indices
train_idx = X_train.index
test_idx = X_test.index

X_train_meta = df.loc[train_idx, categorical_cols + binary_cols].copy()
X_test_meta = df.loc[test_idx, categorical_cols + binary_cols].copy()

# One-hot encode categorical variables
encoder = OneHotEncoder(
    handle_unknown="ignore",
    sparse_output=True
)

X_train_cat = encoder.fit_transform(X_train_meta[categorical_cols])
X_test_cat = encoder.transform(X_test_meta[categorical_cols])

# Binary features
X_train_bin = csr_matrix(
    X_train_meta[binary_cols].astype(float).values
)

X_test_bin = csr_matrix(
    X_test_meta[binary_cols].astype(float).values
)

# Combine categorical + binary features
X_train_meta_encoded = hstack([
    X_train_cat,
    X_train_bin
]).tocsr()

X_test_meta_encoded = hstack([
    X_test_cat,
    X_test_bin
]).tocsr()

print("Metadata train shape:", X_train_meta_encoded.shape)
print("Metadata test shape:", X_test_meta_encoded.shape)

Metadata train shape: (14304, 199)
Metadata test shape: (3576, 199)


In [24]:
X_train_combined = hstack([
    X_train_tfidf,
    X_train_meta_encoded
]).tocsr()

X_test_combined = hstack([
    X_test_tfidf,
    X_test_meta_encoded
]).tocsr()

print("Combined train shape:", X_train_combined.shape)
print("Combined test shape:", X_test_combined.shape)

Combined train shape: (14304, 376489)
Combined test shape: (3576, 376489)


In [25]:
svm_combined = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_combined.fit(X_train_combined, y_train)

y_pred_combined = svm_combined.predict(X_test_combined)
y_score_combined = svm_combined.decision_function(X_test_combined)

In [26]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_auc_score,
    average_precision_score
)

print("Classification Report:")
print(classification_report(y_test, y_pred_combined, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_combined))

print("\nROC-AUC:",
      roc_auc_score(y_test, y_score_combined))

print("PR-AUC:",
      average_precision_score(y_test, y_score_combined))

Classification Report:
              precision    recall  f1-score   support

           0     0.9942    0.9994    0.9968      3403
           1     0.9871    0.8844    0.9329       173

    accuracy                         0.9938      3576
   macro avg     0.9906    0.9419    0.9649      3576
weighted avg     0.9938    0.9938    0.9937      3576


Confusion Matrix:
[[3401    2]
 [  20  153]]

ROC-AUC: 0.9919231416006618
PR-AUC: 0.9652928136632372


In [27]:
import numpy as np

# Get feature names
feature_names = np.array(
    list(tfidf.get_feature_names_out()) +
    list(encoder.get_feature_names_out(categorical_cols)) +
    binary_cols
)

# SVM coefficients
coefficients = svm_combined.coef_[0]

# Features with strongest positive weight → class 1
top_positive_idx = np.argsort(coefficients)[-30:][::-1]

print("Top features associated with class 1 (fraudulent):\n")

for idx in top_positive_idx:
    print(f"{feature_names[idx]:50s} {coefficients[idx]:.4f}")

Top features associated with class 1 (fraudulent):

using link                                         1.3196
line requirements                                  1.2689
money                                              1.2211
unknown apply                                      1.1859
apply using                                        1.1725
unknown                                            1.1413
rohan                                              1.0083
executive sales                                    0.9998
data entry                                         0.9981
earn                                               0.9902
accion                                             0.9351
link                                               0.9239
000                                                0.8994
timejob                                            0.8756
junior project                                     0.8619
high school                                        0.8451
aptitude staffing   

In [28]:
# Text length and word-count features
text_cols = [
    "title",
    "company_profile",
    "description",
    "requirements",
    "benefits"
]

for col in text_cols:
    df[f"{col}_chars"] = df[col].str.len()
    df[f"{col}_words"] = df[col].str.split().str.len()

# Missing-information indicators
for col in ["company_profile", "requirements", "benefits"]:
    df[f"{col}_missing"] = (
        df[col].eq("Unknown").astype(int)
    )

print(df[
    [f"{c}_chars" for c in text_cols] +
    [f"{c}_words" for c in text_cols] +
    ["company_profile_missing",
     "requirements_missing",
     "benefits_missing"]
].head())

   title_chars  company_profile_chars  description_chars  requirements_chars  \
0           16                    885                905                 852   
1           41                   1286               2077                1433   
2           39                    879                355                1363   
3           33                    614               2600                1429   
4           19                   1628               1520                 757   

   benefits_chars  title_words  company_profile_words  description_words  \
0               7            2                    141                124   
1            1292            6                    153                315   
2               7            4                    141                 50   
3             782            5                     85                346   
4              21            3                    207                168   

   requirements_words  benefits_words  company_profile_missing

In [29]:
from scipy.sparse import csr_matrix, hstack

# Numerical engineered features
engineered_cols = (
    [f"{c}_chars" for c in text_cols] +
    [f"{c}_words" for c in text_cols] +
    [
        "company_profile_missing",
        "requirements_missing",
        "benefits_missing"
    ]
)

X_train_eng = csr_matrix(
    df.loc[train_idx, engineered_cols].astype(float).values
)

X_test_eng = csr_matrix(
    df.loc[test_idx, engineered_cols].astype(float).values
)

print("Engineered train shape:", X_train_eng.shape)
print("Engineered test shape:", X_test_eng.shape)

Engineered train shape: (14304, 13)
Engineered test shape: (3576, 13)


In [30]:
X_train_final = hstack([
    X_train_tfidf,
    X_train_meta_encoded,
    X_train_eng
]).tocsr()

X_test_final = hstack([
    X_test_tfidf,
    X_test_meta_encoded,
    X_test_eng
]).tocsr()

print("Final train shape:", X_train_final.shape)
print("Final test shape:", X_test_final.shape)

Final train shape: (14304, 376502)
Final test shape: (3576, 376502)


In [31]:
svm_final = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_final.fit(X_train_final, y_train)

y_pred_final = svm_final.predict(X_test_final)
y_score_final = svm_final.decision_function(X_test_final)

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [32]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_auc_score,
    average_precision_score
)

print("Classification Report:")
print(classification_report(y_test, y_pred_final, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_final))

print("\nROC-AUC:",
      roc_auc_score(y_test, y_score_final))

print("PR-AUC:",
      average_precision_score(y_test, y_score_final))

Classification Report:
              precision    recall  f1-score   support

           0     0.9536    0.9545    0.9540      3403
           1     0.0882    0.0867    0.0875       173

    accuracy                         0.9125      3576
   macro avg     0.5209    0.5206    0.5207      3576
weighted avg     0.9117    0.9125    0.9121      3576


Confusion Matrix:
[[3248  155]
 [ 158   15]]

ROC-AUC: 0.6371528691956603
PR-AUC: 0.07809354381222641


In [33]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_eng_scaled = scaler.fit_transform(
    X_train_eng.toarray()
)

X_test_eng_scaled = scaler.transform(
    X_test_eng.toarray()
)

X_train_eng_scaled = csr_matrix(X_train_eng_scaled)
X_test_eng_scaled = csr_matrix(X_test_eng_scaled)

print(X_train_eng_scaled.shape)
print(X_test_eng_scaled.shape)

(14304, 13)
(3576, 13)


In [34]:
X_train_scaled = hstack([
    X_train_tfidf,
    X_train_meta_encoded,
    X_train_eng_scaled
]).tocsr()

X_test_scaled = hstack([
    X_test_tfidf,
    X_test_meta_encoded,
    X_test_eng_scaled
]).tocsr()

print("Train:", X_train_scaled.shape)
print("Test:", X_test_scaled.shape)

Train: (14304, 376502)
Test: (3576, 376502)


In [35]:
svm_scaled = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = svm_scaled.predict(X_test_scaled)
y_score_scaled = svm_scaled.decision_function(X_test_scaled)

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [36]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    roc_auc_score,
    average_precision_score
)

print("Classification Report:")
print(classification_report(y_test, y_pred_scaled, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_scaled))

print("\nROC-AUC:",
      roc_auc_score(y_test, y_score_scaled))

print("PR-AUC:",
      average_precision_score(y_test, y_score_scaled))

Classification Report:
              precision    recall  f1-score   support

           0     0.9939    0.9988    0.9963      3403
           1     0.9744    0.8786    0.9240       173

    accuracy                         0.9930      3576
   macro avg     0.9841    0.9387    0.9602      3576
weighted avg     0.9929    0.9930    0.9928      3576


Confusion Matrix:
[[3399    4]
 [  21  152]]

ROC-AUC: 0.992738471155169
PR-AUC: 0.963639463960177


In [37]:
# Find false negatives
fn_mask = (y_test == 1) & (y_pred_combined == 0)

false_negatives = df.loc[
    y_test.index[fn_mask],
    [
        "title",
        "location",
        "employment_type",
        "required_experience",
        "required_education",
        "industry",
        "function",
        "telecommuting",
        "has_company_logo",
        "has_questions",
        "company_profile",
        "description",
        "requirements",
        "benefits"
    ]
].copy()

print("False negatives:", len(false_negatives))

false_negatives[
    [
        "title",
        "location",
        "employment_type",
        "required_experience",
        "required_education",
        "industry",
        "function",
        "telecommuting",
        "has_company_logo",
        "has_questions"
    ]
]

False negatives: 20


,title,location,employment_type,required_experience,required_education,industry,function,telecommuting,has_company_logo,has_questions
17648,Vemma Brand Partner,"US, CA, 91401",Full-time,Not Applicable,Unknown,Marketing and Advertising,Marketing,0,0,1
12104,Production Manager Heavy Duty Diesel 2022,"US, TN, Morristown",Full-time,Unknown,Unknown,Unknown,Unknown,0,0,0
17598,Real Estate & Insurance Professionals,"US, GA, Atlanta",Unknown,Unknown,Unknown,Unknown,Unknown,0,0,1
10407,Hotel Workers Needed,GB,Full-time,Not Applicable,High School or equivalent,Hospitality,Customer Service,0,1,0
17808,"vacancies,nannies,cooks,drivers and housekeepe...","US, FL,",Full-time,Not Applicable,Unspecified,Hospital & Health Care,Health Care Provider,0,0,1
5926,Assistant Accountant,"AU, ,",Part-time,Unknown,Unknown,Financial Services,Finance,0,0,0
3463,Business Development Manager,"US, CA, San Francisco",Full-time,Unknown,High School or equivalent,Wholesale,Sales,0,0,1
6144,Physician Assistant PA Dermatology 1843,"US, NY, Ridgewood",Full-time,Not Applicable,Bachelor's Degree,"Health, Wellness and Fitness",Unknown,0,0,0
17554,Entry Level Medical Transcriptionist,"US, CA, Los Angeles",Full-time,Entry level,High School or equivalent,Medical Practice,Unknown,1,0,0
17644,Fidelity,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,0,0,0


In [38]:
import re

df["has_money_language"] = df["combined_text"].str.contains(
    r"\bmoney\b|\bearn\b|\bincome\b|\bprofit\b|\bwealth\b|\bpay\b",
    case=False,
    regex=True
).astype(int)

df["has_easy_money_language"] = df["combined_text"].str.contains(
    r"easy money|make money|earn money|make extra money|work from home",
    case=False,
    regex=True
).astype(int)

df["has_immediate_language"] = df["combined_text"].str.contains(
    r"immediate start|start immediately|urgent|hiring immediately",
    case=False,
    regex=True
).astype(int)

df["has_application_link"] = df["combined_text"].str.contains(
    r"https?://|www\.|apply.*link|using.*link",
    case=False,
    regex=True
).astype(int)

df["has_recruitment_language"] = df["combined_text"].str.contains(
    r"brand partner|representative needed|work from home|no experience",
    case=False,
    regex=True
).astype(int)

new_feature_cols = [
    "has_money_language",
    "has_easy_money_language",
    "has_immediate_language",
    "has_application_link",
    "has_recruitment_language"
]

print(df[new_feature_cols].sum())

has_money_language          3366
has_easy_money_language      368
has_immediate_language       530
has_application_link         619
has_recruitment_language     456
dtype: int64


In [39]:
for col in new_feature_cols:
    print(f"\n{col}")
    print(
        df.groupby(col)["fraudulent"]
          .agg(["count", "sum", "mean"])
    )


has_money_language
                    count  sum      mean
has_money_language                      
0                   14514  541  0.037274
1                    3366  325  0.096554

has_easy_money_language
                         count  sum      mean
has_easy_money_language                      
0                        17512  760  0.043399
1                          368  106  0.288043

has_immediate_language
                        count  sum      mean
has_immediate_language                      
0                       17350  798  0.045994
1                         530   68  0.128302

has_application_link
                      count  sum      mean
has_application_link                      
0                     17261  804  0.046579
1                       619   62  0.100162

has_recruitment_language
                          count  sum      mean
has_recruitment_language                      
0                         17424  730  0.041896
1                           456  136  0.29

In [40]:
X_train_phrase = csr_matrix(
    df.loc[train_idx, new_feature_cols].astype(float).values
)

X_test_phrase = csr_matrix(
    df.loc[test_idx, new_feature_cols].astype(float).values
)

print("Phrase train:", X_train_phrase.shape)
print("Phrase test:", X_test_phrase.shape)

Phrase train: (14304, 5)
Phrase test: (3576, 5)


In [41]:
X_train_phrase_model = hstack([
    X_train_tfidf,
    X_train_meta_encoded,
    X_train_phrase
]).tocsr()

X_test_phrase_model = hstack([
    X_test_tfidf,
    X_test_meta_encoded,
    X_test_phrase
]).tocsr()

print("Train:", X_train_phrase_model.shape)
print("Test:", X_test_phrase_model.shape)

Train: (14304, 376494)
Test: (3576, 376494)


In [42]:
svm_phrase = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_phrase.fit(X_train_phrase_model, y_train)

y_pred_phrase = svm_phrase.predict(X_test_phrase_model)
y_score_phrase = svm_phrase.decision_function(X_test_phrase_model)

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [43]:
print("Classification Report:")
print(classification_report(y_test, y_pred_phrase, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_phrase))

print("\nROC-AUC:",
      roc_auc_score(y_test, y_score_phrase))

print("PR-AUC:",
      average_precision_score(y_test, y_score_phrase))

Classification Report:
              precision    recall  f1-score   support

           0     0.9947    0.9988    0.9968      3403
           1     0.9748    0.8960    0.9337       173

    accuracy                         0.9938      3576
   macro avg     0.9848    0.9474    0.9653      3576
weighted avg     0.9938    0.9938    0.9937      3576


Confusion Matrix:
[[3399    4]
 [  18  155]]

ROC-AUC: 0.9918246226128256
PR-AUC: 0.9661572674735024


In [44]:
from sklearn.metrics import precision_score, recall_score, f1_score

thresholds = [-0.5, -0.3, -0.2, -0.1, 0, 0.1, 0.2, 0.3, 0.5]

for threshold in thresholds:
    y_pred_threshold = (y_score_phrase >= threshold).astype(int)

    precision = precision_score(y_test, y_pred_threshold)
    recall = recall_score(y_test, y_pred_threshold)
    f1 = f1_score(y_test, y_pred_threshold)

    print(
        f"Threshold: {threshold:>4} | "
        f"Precision: {precision:.4f} | "
        f"Recall: {recall:.4f} | "
        f"F1: {f1:.4f}"
    )

Threshold: -0.5 | Precision: 0.7174 | Recall: 0.9538 | F1: 0.8189
Threshold: -0.3 | Precision: 0.8859 | Recall: 0.9422 | F1: 0.9132
Threshold: -0.2 | Precision: 0.9249 | Recall: 0.9249 | F1: 0.9249
Threshold: -0.1 | Precision: 0.9464 | Recall: 0.9191 | F1: 0.9326
Threshold:    0 | Precision: 0.9748 | Recall: 0.8960 | F1: 0.9337
Threshold:  0.1 | Precision: 0.9869 | Recall: 0.8728 | F1: 0.9264
Threshold:  0.2 | Precision: 1.0000 | Recall: 0.8208 | F1: 0.9016
Threshold:  0.3 | Precision: 1.0000 | Recall: 0.7977 | F1: 0.8875
Threshold:  0.5 | Precision: 1.0000 | Recall: 0.7283 | F1: 0.8428


In [45]:
import numpy as np
from sklearn.metrics import precision_score, recall_score, f1_score

thresholds = np.arange(-0.30, 0.11, 0.01)

results = []

for threshold in thresholds:
    y_pred = (y_score_phrase >= threshold).astype(int)

    results.append({
        "threshold": threshold,
        "precision": precision_score(y_test, y_pred),
        "recall": recall_score(y_test, y_pred),
        "f1": f1_score(y_test, y_pred)
    })

threshold_df = pd.DataFrame(results)

print(threshold_df.sort_values("f1", ascending=False).head(10))

       threshold  precision    recall        f1
33  3.000000e-02   0.981013  0.895954  0.936556
32  2.000000e-02   0.981013  0.895954  0.936556
35  5.000000e-02   0.981013  0.895954  0.936556
34  4.000000e-02   0.981013  0.895954  0.936556
28 -2.000000e-02   0.968944  0.901734  0.934132
27 -3.000000e-02   0.968944  0.901734  0.934132
29 -1.000000e-02   0.974843  0.895954  0.933735
30  2.775558e-16   0.974843  0.895954  0.933735
31  1.000000e-02   0.974843  0.895954  0.933735
14 -1.600000e-01   0.941176  0.924855  0.932945


In [46]:
best_threshold = 0.03

y_pred_final = (y_score_phrase >= best_threshold).astype(int)

print(classification_report(y_test, y_pred_final, digits=4))
print(confusion_matrix(y_test, y_pred_final))

print("ROC-AUC:", roc_auc_score(y_test, y_score_phrase))
print("PR-AUC:", average_precision_score(y_test, y_score_phrase))

              precision    recall  f1-score   support

           0     0.9947    0.9991    0.9969      3403
           1     0.9810    0.8960    0.9366       173

    accuracy                         0.9941      3576
   macro avg     0.9879    0.9475    0.9667      3576
weighted avg     0.9941    0.9941    0.9940      3576

[[3400    3]
 [  18  155]]
ROC-AUC: 0.9918246226128256
PR-AUC: 0.9661572674735024


In [47]:
weights = [
    {0: 1, 1: 1.5},
    {0: 1, 1: 2},
    {0: 1, 1: 2.5},
    {0: 1, 1: 3}
]

for weight in weights:
    model = LinearSVC(
        class_weight=weight,
        C=1.0,
        random_state=42
    )

    model.fit(X_train_phrase_model, y_train)

    pred = model.predict(X_test_phrase_model)
    score = model.decision_function(X_test_phrase_model)

    print("\nClass weight:", weight)
    print("Precision:", precision_score(y_test, pred))
    print("Recall:", recall_score(y_test, pred))
    print("F1:", f1_score(y_test, pred))


Class weight: {0: 1, 1: 1.5}
Precision: 1.0
Recall: 0.8265895953757225
F1: 0.9050632911392406

Class weight: {0: 1, 1: 2}
Precision: 1.0
Recall: 0.8439306358381503
F1: 0.9153605015673981

Class weight: {0: 1, 1: 2.5}
Precision: 1.0
Recall: 0.861271676300578
F1: 0.9254658385093167

Class weight: {0: 1, 1: 3}
Precision: 0.9933774834437086
Recall: 0.8670520231213873
F1: 0.9259259259259259


In [48]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(
    class_weight="balanced",
    C=1.0,
    max_iter=2000,
    random_state=42
)

lr.fit(X_train_phrase_model, y_train)

y_pred_lr = lr.predict(X_test_phrase_model)
y_prob_lr = lr.predict_proba(X_test_phrase_model)[:, 1]

print(classification_report(y_test, y_pred_lr, digits=4))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_lr))

print("\nROC-AUC:", roc_auc_score(y_test, y_prob_lr))
print("PR-AUC:", average_precision_score(y_test, y_prob_lr))

              precision    recall  f1-score   support

           0     0.9961    0.9821    0.9891      3403
           1     0.7240    0.9249    0.8122       173

    accuracy                         0.9793      3576
   macro avg     0.8601    0.9535    0.9006      3576
weighted avg     0.9830    0.9793    0.9805      3576


Confusion Matrix:
[[3342   61]
 [  13  160]]

ROC-AUC: 0.9918365128354953
PR-AUC: 0.934369762286639


In [49]:
thresholds = np.arange(0.10, 0.91, 0.01)

lr_results = []

for threshold in thresholds:
    pred = (y_prob_lr >= threshold).astype(int)

    lr_results.append({
        "threshold": threshold,
        "precision": precision_score(y_test, pred),
        "recall": recall_score(y_test, pred),
        "f1": f1_score(y_test, pred)
    })

lr_threshold_df = pd.DataFrame(lr_results)

print(
    lr_threshold_df
    .sort_values("f1", ascending=False)
    .head(10)
)

    threshold  precision    recall        f1
57       0.67   0.877193  0.867052  0.872093
58       0.68   0.881657  0.861272  0.871345
56       0.66   0.867816  0.872832  0.870317
54       0.64   0.858757  0.878613  0.868571
55       0.65   0.858757  0.878613  0.868571
59       0.69   0.880952  0.855491  0.868035
60       0.70   0.880952  0.855491  0.868035
61       0.71   0.885542  0.849711  0.867257
53       0.63   0.850000  0.884393  0.866856
51       0.61   0.829787  0.901734  0.864266


In [50]:
from xgboost import XGBClassifier

xgb = XGBClassifier(
    n_estimators=300,
    max_depth=6,
    learning_rate=0.05,
    subsample=0.8,
    colsample_bytree=0.8,
    objective="binary:logistic",
    eval_metric="logloss",
    random_state=42,
    n_jobs=-1
)

xgb.fit(
    X_train_meta_encoded,
    y_train
)

y_pred_xgb = xgb.predict(X_test_meta_encoded)
y_prob_xgb = xgb.predict_proba(X_test_meta_encoded)[:, 1]

print(classification_report(y_test, y_pred_xgb, digits=4))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_xgb))

print("\nROC-AUC:", roc_auc_score(y_test, y_prob_xgb))
print("PR-AUC:", average_precision_score(y_test, y_prob_xgb))

              precision    recall  f1-score   support

           0     0.9712    0.9991    0.9849      3403
           1     0.9600    0.4162    0.5806       173

    accuracy                         0.9709      3576
   macro avg     0.9656    0.7077    0.7828      3576
weighted avg     0.9706    0.9709    0.9654      3576


Confusion Matrix:
[[3400    3]
 [ 101   72]]

ROC-AUC: 0.9415145425916269
PR-AUC: 0.6802034897834436


In [51]:
from sklearn.naive_bayes import ComplementNB

cnb = ComplementNB(alpha=1.0)

cnb.fit(X_train_tfidf, y_train)

y_pred_cnb = cnb.predict(X_test_tfidf)
y_prob_cnb = cnb.predict_proba(X_test_tfidf)[:, 1]

print(classification_report(y_test, y_pred_cnb, digits=4))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_cnb))

print("\nROC-AUC:", roc_auc_score(y_test, y_prob_cnb))
print("PR-AUC:", average_precision_score(y_test, y_prob_cnb))

              precision    recall  f1-score   support

           0     0.9629    0.9982    0.9802      3403
           1     0.8750    0.2428    0.3801       173

    accuracy                         0.9617      3576
   macro avg     0.9189    0.6205    0.6802      3576
weighted avg     0.9586    0.9617    0.9512      3576


Confusion Matrix:
[[3397    6]
 [ 131   42]]

ROC-AUC: 0.9324533436155449
PR-AUC: 0.602025194854186


In [52]:
fn_mask = (y_test == 1) & (y_pred_final == 0)

fn_indices = y_test[fn_mask].index

fn_df = df.loc[
    fn_indices,
    ["job_id", "title", "location", "employment_type",
     "required_experience", "required_education",
     "industry", "function", "telecommuting",
     "has_company_logo", "has_questions", "fraudulent"]
]

print("False Negatives:", len(fn_df))
display(fn_df)

False Negatives: 18


,job_id,title,location,employment_type,required_experience,required_education,industry,function,telecommuting,has_company_logo,has_questions,fraudulent
17648,17649,Vemma Brand Partner,"US, CA, 91401",Full-time,Not Applicable,Unknown,Marketing and Advertising,Marketing,0,0,1,1
12104,12105,Production Manager Heavy Duty Diesel 2022,"US, TN, Morristown",Full-time,Unknown,Unknown,Unknown,Unknown,0,0,0,1
17598,17599,Real Estate & Insurance Professionals,"US, GA, Atlanta",Unknown,Unknown,Unknown,Unknown,Unknown,0,0,1,1
10407,10408,Hotel Workers Needed,GB,Full-time,Not Applicable,High School or equivalent,Hospitality,Customer Service,0,1,0,1
17808,17809,"vacancies,nannies,cooks,drivers and housekeepe...","US, FL,",Full-time,Not Applicable,Unspecified,Hospital & Health Care,Health Care Provider,0,0,1,1
5926,5927,Assistant Accountant,"AU, ,",Part-time,Unknown,Unknown,Financial Services,Finance,0,0,0,1
3463,3464,Business Development Manager,"US, CA, San Francisco",Full-time,Unknown,High School or equivalent,Wholesale,Sales,0,0,1,1
6144,6145,Physician Assistant PA Dermatology 1843,"US, NY, Ridgewood",Full-time,Not Applicable,Bachelor's Degree,"Health, Wellness and Fitness",Unknown,0,0,0,1
17554,17555,Entry Level Medical Transcriptionist,"US, CA, Los Angeles",Full-time,Entry level,High School or equivalent,Medical Practice,Unknown,1,0,0,1
17644,17645,Fidelity,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,0,0,0,1


In [53]:
from sklearn.feature_extraction.text import TfidfVectorizer

char_tfidf = TfidfVectorizer(
    analyzer="char",
    ngram_range=(3, 5),
    min_df=3,
    max_features=200000,
    sublinear_tf=True
)

X_train_char = char_tfidf.fit_transform(X_train)
X_test_char = char_tfidf.transform(X_test)

print("Train:", X_train_char.shape)
print("Test:", X_test_char.shape)

Train: (14304, 200000)
Test: (3576, 200000)


In [54]:
X_train_char_combined = hstack([
    X_train_tfidf,
    X_train_meta_encoded,
    X_train_phrase,
    X_train_char
]).tocsr()

X_test_char_combined = hstack([
    X_test_tfidf,
    X_test_meta_encoded,
    X_test_phrase,
    X_test_char
]).tocsr()

print("Train:", X_train_char_combined.shape)
print("Test:", X_test_char_combined.shape)

Train: (14304, 576494)
Test: (3576, 576494)


In [55]:
svm_char = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_char.fit(X_train_char_combined, y_train)

y_pred_char = svm_char.predict(X_test_char_combined)
y_score_char = svm_char.decision_function(X_test_char_combined)

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [56]:
print("Classification Report:")
print(classification_report(y_test, y_pred_char, digits=4))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_char))

print("\nROC-AUC:", roc_auc_score(y_test, y_score_char))
print("PR-AUC:", average_precision_score(y_test, y_score_char))

Classification Report:
              precision    recall  f1-score   support

           0     0.9939    0.9988    0.9963      3403
           1     0.9744    0.8786    0.9240       173

    accuracy                         0.9930      3576
   macro avg     0.9841    0.9387    0.9602      3576
weighted avg     0.9929    0.9930    0.9928      3576


Confusion Matrix:
[[3399    4]
 [  21  152]]

ROC-AUC: 0.9911214008720628
PR-AUC: 0.9684207188958173


In [57]:
thresholds = np.arange(-0.30, 0.11, 0.01)

char_results = []

for threshold in thresholds:
    pred = (y_score_char >= threshold).astype(int)

    char_results.append({
        "threshold": threshold,
        "precision": precision_score(y_test, pred),
        "recall": recall_score(y_test, pred),
        "f1": f1_score(y_test, pred)
    })

char_threshold_df = pd.DataFrame(char_results)

print(
    char_threshold_df
    .sort_values("f1", ascending=False)
    .head(10)
)

    threshold  precision    recall        f1
12      -0.18   0.958333  0.930636  0.944282
11      -0.19   0.958333  0.930636  0.944282
13      -0.17   0.958333  0.930636  0.944282
10      -0.20   0.952663  0.930636  0.941520
9       -0.21   0.952663  0.930636  0.941520
14      -0.16   0.958084  0.924855  0.941176
16      -0.14   0.963636  0.919075  0.940828
20      -0.10   0.969325  0.913295  0.940476
19      -0.11   0.969325  0.913295  0.940476
8       -0.22   0.947059  0.930636  0.938776


In [58]:
best_char_threshold = -0.18

y_pred_char_final = (
    y_score_char >= best_char_threshold
).astype(int)

print(classification_report(
    y_test,
    y_pred_char_final,
    digits=4
))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_char_final))

print("\nROC-AUC:", roc_auc_score(y_test, y_score_char))
print("PR-AUC:", average_precision_score(y_test, y_score_char))

              precision    recall  f1-score   support

           0     0.9965    0.9979    0.9972      3403
           1     0.9583    0.9306    0.9443       173

    accuracy                         0.9947      3576
   macro avg     0.9774    0.9643    0.9707      3576
weighted avg     0.9946    0.9947    0.9946      3576


Confusion Matrix:
[[3396    7]
 [  12  161]]

ROC-AUC: 0.9911214008720628
PR-AUC: 0.9684207188958173


In [59]:
fn_mask_char = (y_test == 1) & (y_pred_char_final == 0)

fn_indices_char = y_test[fn_mask_char].index

fn_char_df = df.loc[
    fn_indices_char,
    [
        "job_id", "title", "location",
        "employment_type", "required_experience",
        "required_education", "industry", "function",
        "telecommuting", "has_company_logo",
        "has_questions", "fraudulent"
    ]
]

print("False Negatives:", len(fn_char_df))
display(fn_char_df)

False Negatives: 12


,job_id,title,location,employment_type,required_experience,required_education,industry,function,telecommuting,has_company_logo,has_questions,fraudulent
17648,17649,Vemma Brand Partner,"US, CA, 91401",Full-time,Not Applicable,Unknown,Marketing and Advertising,Marketing,0,0,1,1
17598,17599,Real Estate & Insurance Professionals,"US, GA, Atlanta",Unknown,Unknown,Unknown,Unknown,Unknown,0,0,1,1
17808,17809,"vacancies,nannies,cooks,drivers and housekeepe...","US, FL,",Full-time,Not Applicable,Unspecified,Hospital & Health Care,Health Care Provider,0,0,1,1
5926,5927,Assistant Accountant,"AU, ,",Part-time,Unknown,Unknown,Financial Services,Finance,0,0,0,1
3463,3464,Business Development Manager,"US, CA, San Francisco",Full-time,Unknown,High School or equivalent,Wholesale,Sales,0,0,1,1
17554,17555,Entry Level Medical Transcriptionist,"US, CA, Los Angeles",Full-time,Entry level,High School or equivalent,Medical Practice,Unknown,1,0,0,1
17644,17645,Fidelity,Unknown,Unknown,Unknown,Unknown,Unknown,Unknown,0,0,0,1
5672,5673,Software Design Engineer,"US, CA, Palo Alto",Full-time,Mid-Senior level,Bachelor's Degree,Computer Software,Engineering,0,0,1,1
4761,4762,COMPANY REPRESENTATIVE NEEDED,"US, NY, mumford",Contract,Not Applicable,High School or equivalent,Civic & Social Organization,Finance,0,0,1,1
5499,5500,Brand Partner,"US, CA,",Unknown,Unknown,Unknown,Unknown,Unknown,0,0,1,1


In [60]:
for idx in fn_indices_char:
    row = df.loc[idx]

    print("=" * 100)
    print("TITLE:", row["title"])
    print("COMPANY:", row["company_profile"])
    print("DESCRIPTION:", row["description"])
    print("REQUIREMENTS:", row["requirements"])
    print("BENEFITS:", row["benefits"])
    print()

TITLE: Vemma Brand Partner
COMPANY: Unknown
DESCRIPTION: Looking for motivated and hard working individuals!*Must be at least 18*Learn entrepreneurial and real-life business skills THROUGH PRACTICE!Study Success and Follow successful people!Vemma is a health and nutrition company on the rise!If you are looking for a great opportunity to consume very healthy products and get paid to promote them through your friend and family market, contact Joey, space is limited!
REQUIREMENTS: Motivated and willing to work with a team!
BENEFITS: Unknown

TITLE: Real Estate & Insurance Professionals
COMPANY: Elite Realty Services was formed in 2004 by its Founder, Robert Do, who originally owned and operated several Coldwell Banker offices.  Realizing the Franchise had some limitations in their services to the public and to its agents, he decided to form Elite Realty Services. His vision for the new company was to offer more to meet the needs of the public and meeting those needs, he knew the agents ne

In [61]:
import re
import numpy as np

def count_pattern(text, pattern):
    return len(re.findall(pattern, str(text), flags=re.IGNORECASE))

df["url_count"] = df["combined_text"].apply(
    lambda x: count_pattern(x, r"https?://|www\.|#URL_")
)

df["email_count"] = df["combined_text"].apply(
    lambda x: count_pattern(x, r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b|#EMAIL_")
)

df["phone_count"] = df["combined_text"].apply(
    lambda x: count_pattern(x, r"\+?\d[\d\s().-]{7,}\d|#PHONE_")
)

df["money_symbol_count"] = df["combined_text"].apply(
    lambda x: count_pattern(x, r"[$£€]")
)

df["exclamation_count"] = df["combined_text"].str.count("!")

df["all_caps_count"] = df["combined_text"].apply(
    lambda x: len(re.findall(r"\b[A-Z]{3,}\b", str(x)))
)

df["multiple_role_count"] = df["title"].apply(
    lambda x: len(re.split(r",|/| and ", str(x), flags=re.IGNORECASE))
)

df["text_length"] = df["combined_text"].str.len()

df["word_count"] = df["combined_text"].str.split().str.len()

In [62]:
new_structural_cols = [
    "url_count",
    "email_count",
    "phone_count",
    "money_symbol_count",
    "exclamation_count",
    "all_caps_count",
    "multiple_role_count",
    "text_length",
    "word_count"
]

print(df.groupby("fraudulent")[new_structural_cols].mean())

            url_count  email_count  phone_count  money_symbol_count  \
fraudulent                                                            
0            0.629540     0.072881     0.348654            0.326731   
1            0.498845     0.257506     0.354503            0.784065   

            exclamation_count  all_caps_count  multiple_role_count  \
fraudulent                                                           
0                    0.668097        4.342835             1.194369   
1                    0.554273        5.518476             1.289838   

            text_length  word_count  
fraudulent                           
0           2705.570354  380.202892  
1           2087.578522  283.609700  


In [63]:
structural_cols = [
    "email_count",
    "money_symbol_count",
    "all_caps_count",
    "text_length",
    "word_count",
]

In [64]:
print(
    df[
        [
            "email_count",
            "money_symbol_count",
            "all_caps_count",
            "text_length",
            "word_count"
        ]
    ].corr().round(3)
)

                    email_count  money_symbol_count  all_caps_count  \
email_count               1.000              -0.007           0.090   
money_symbol_count       -0.007               1.000           0.122   
all_caps_count            0.090               0.122           1.000   
text_length               0.078              -0.061           0.280   
word_count                0.060              -0.051           0.261   

                    text_length  word_count  
email_count               0.078       0.060  
money_symbol_count       -0.061      -0.051  
all_caps_count            0.280       0.261  
text_length               1.000       0.973  
word_count                0.973       1.000  


In [65]:
structural_cols = [
    "email_count",
    "money_symbol_count",
    "all_caps_count",
    "word_count"
]

In [66]:
from sklearn.preprocessing import StandardScaler
from scipy.sparse import hstack, csr_matrix

scaler_struct = StandardScaler()

X_train_struct = scaler_struct.fit_transform(
    df.loc[train_idx, structural_cols].astype(float)
)

X_test_struct = scaler_struct.transform(
    df.loc[test_idx, structural_cols].astype(float)
)

X_train_struct = csr_matrix(X_train_struct)
X_test_struct = csr_matrix(X_test_struct)

print(X_train_struct.shape)
print(X_test_struct.shape)

(14304, 4)
(3576, 4)


In [67]:
X_train_struct_model = hstack([
    X_train_char_combined,
    X_train_struct
]).tocsr()

X_test_struct_model = hstack([
    X_test_char_combined,
    X_test_struct
]).tocsr()

print(X_train_struct_model.shape)
print(X_test_struct_model.shape)

(14304, 576498)
(3576, 576498)


In [68]:
svm_struct = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_struct.fit(X_train_struct_model, y_train)

y_score_struct = svm_struct.decision_function(X_test_struct_model)

y_pred_struct = (
    y_score_struct >= -0.18
).astype(int)

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [69]:
print(classification_report(
    y_test,
    y_pred_struct,
    digits=4
))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_struct))

print("\nROC-AUC:", roc_auc_score(y_test, y_score_struct))
print("PR-AUC:", average_precision_score(y_test, y_score_struct))

              precision    recall  f1-score   support

           0     0.9959    0.9979    0.9969      3403
           1     0.9578    0.9191    0.9381       173

    accuracy                         0.9941      3576
   macro avg     0.9769    0.9585    0.9675      3576
weighted avg     0.9941    0.9941    0.9941      3576


Confusion Matrix:
[[3396    7]
 [  14  159]]

ROC-AUC: 0.9895705761152604
PR-AUC: 0.967842146647475


In [70]:
from sklearn.model_selection import train_test_split

X_tr, X_val, y_tr, y_val = train_test_split(
    X_train_char_combined,
    y_train,
    test_size=0.20,
    random_state=42,
    stratify=y_train
)

print("Train:", X_tr.shape)
print("Validation:", X_val.shape)
print("Test:", X_test_char_combined.shape)

Train: (11443, 576494)
Validation: (2861, 576494)
Test: (3576, 576494)


In [71]:
svm_final = LinearSVC(
    class_weight="balanced",
    C=1.0,
    random_state=42
)

svm_final.fit(X_tr, y_tr)

y_val_score = svm_final.decision_function(X_val)

print("Validation scores generated:", len(y_val_score))

Validation scores generated: 2861


In [73]:
from sklearn.metrics import f1_score, precision_score, recall_score

thresholds = np.arange(-0.50, 0.51, 0.01)

results = []

for threshold in thresholds:
    y_val_pred = (y_val_score >= threshold).astype(int)

    results.append({
        "threshold": threshold,
        "precision": precision_score(y_val, y_val_pred, zero_division=0),
        "recall": recall_score(y_val, y_val_pred, zero_division=0),
        "f1": f1_score(y_val, y_val_pred, zero_division=0)
    })

threshold_df = pd.DataFrame(results)

best_row = threshold_df.loc[
    threshold_df["f1"].idxmax()
]

print(best_row)

from sklearn.metrics import f1_score, precision_score, recall_score

thresholds = np.arange(-0.50, 0.51, 0.01)

results = []

for threshold in thresholds:
    y_val_pred = (y_val_score >= threshold).astype(int)

    results.append({
        "threshold": threshold,
        "precision": precision_score(y_val, y_val_pred, zero_division=0),
        "recall": recall_score(y_val, y_val_pred, zero_division=0),
        "f1": f1_score(y_val, y_val_pred, zero_division=0)
    })

threshold_df = pd.DataFrame(results)

best_row = threshold_df.loc[
    threshold_df["f1"].idxmax()
]

print(best_row)

threshold   -0.270000
precision    0.887324
recall       0.906475
f1           0.896797
Name: 23, dtype: float64
threshold   -0.270000
precision    0.887324
recall       0.906475
f1           0.896797
Name: 23, dtype: float64


In [74]:
best_threshold = -0.27

y_test_score_final = svm_final.decision_function(X_test_char_combined)

y_test_pred_final = (
    y_test_score_final >= best_threshold
).astype(int)

print(classification_report(
    y_test,
    y_test_pred_final,
    digits=4
))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_test_pred_final))

print("\nROC-AUC:", roc_auc_score(y_test, y_test_score_final))
print("PR-AUC:", average_precision_score(y_test, y_test_score_final))

              precision    recall  f1-score   support

           0     0.9962    0.9968    0.9965      3403
           1     0.9357    0.9249    0.9302       173

    accuracy                         0.9933      3576
   macro avg     0.9659    0.9608    0.9634      3576
weighted avg     0.9933    0.9933    0.9933      3576


Confusion Matrix:
[[3392   11]
 [  13  160]]

ROC-AUC: 0.9907850774308287
PR-AUC: 0.9678622924312753


In [75]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    average_precision_score
)

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

cv_results = []

for fold, (train_idx_cv, val_idx_cv) in enumerate(
    skf.split(X_train_char_combined, y_train), 1
):
    print(f"\n===== Fold {fold} =====")

    X_cv_train = X_train_char_combined[train_idx_cv]
    X_cv_val = X_train_char_combined[val_idx_cv]

    y_cv_train = y_train.iloc[train_idx_cv]
    y_cv_val = y_train.iloc[val_idx_cv]

    model_cv = LinearSVC(
        class_weight="balanced",
        C=1.0,
        random_state=42
    )

    model_cv.fit(X_cv_train, y_cv_train)

    y_cv_score = model_cv.decision_function(X_cv_val)

    y_cv_pred = (
        y_cv_score >= -0.27
    ).astype(int)

    fold_result = {
        "fold": fold,
        "accuracy": accuracy_score(y_cv_val, y_cv_pred),
        "precision": precision_score(y_cv_val, y_cv_pred),
        "recall": recall_score(y_cv_val, y_cv_pred),
        "f1": f1_score(y_cv_val, y_cv_pred),
        "roc_auc": roc_auc_score(y_cv_val, y_cv_score),
        "pr_auc": average_precision_score(y_cv_val, y_cv_score)
    }

    cv_results.append(fold_result)

    print(f"Accuracy : {fold_result['accuracy']:.4f}")
    print(f"Precision: {fold_result['precision']:.4f}")
    print(f"Recall   : {fold_result['recall']:.4f}")
    print(f"F1       : {fold_result['f1']:.4f}")
    print(f"ROC-AUC  : {fold_result['roc_auc']:.4f}")
    print(f"PR-AUC   : {fold_result['pr_auc']:.4f}")


===== Fold 1 =====


/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy : 0.9906
Precision: 0.8828
Recall   : 0.9275
F1       : 0.9046
ROC-AUC  : 0.9958
PR-AUC   : 0.9643

===== Fold 2 =====
Accuracy : 0.9860
Precision: 0.8278
Recall   : 0.8993
F1       : 0.8621
ROC-AUC  : 0.9955
PR-AUC   : 0.9155

===== Fold 3 =====
Accuracy : 0.9857
Precision: 0.8403
Recall   : 0.8705
F1       : 0.8551
ROC-AUC  : 0.9906
PR-AUC   : 0.9360

===== Fold 4 =====


/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy : 0.9878
Precision: 0.8662
Recall   : 0.8849
F1       : 0.8754
ROC-AUC  : 0.9922
PR-AUC   : 0.9432

===== Fold 5 =====
Accuracy : 0.9839
Precision: 0.8194
Recall   : 0.8551
F1       : 0.8369
ROC-AUC  : 0.9917
PR-AUC   : 0.9289


In [76]:
LinearSVC(
    class_weight="balanced",
    C=1.0,
    max_iter=5000,
    random_state=42
)

LinearSVC(class_weight='balanced', max_iter=5000, random_state=42)

In [77]:
from sklearn.model_selection import StratifiedKFold
from sklearn.svm import LinearSVC
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    average_precision_score
)

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

cv_results_fixed = []

for fold, (train_idx_cv, val_idx_cv) in enumerate(
    skf.split(X_train_char_combined, y_train), 1
):
    print(f"\n===== Fold {fold} =====")

    X_cv_train = X_train_char_combined[train_idx_cv]
    X_cv_val = X_train_char_combined[val_idx_cv]

    y_cv_train = y_train.iloc[train_idx_cv]
    y_cv_val = y_train.iloc[val_idx_cv]

    model_cv = LinearSVC(
        class_weight="balanced",
        C=1.0,
        max_iter=5000,
        random_state=42
    )

    model_cv.fit(X_cv_train, y_cv_train)

    y_cv_score = model_cv.decision_function(X_cv_val)

    y_cv_pred = (y_cv_score >= -0.27).astype(int)

    fold_result = {
        "fold": fold,
        "accuracy": accuracy_score(y_cv_val, y_cv_pred),
        "precision": precision_score(y_cv_val, y_cv_pred),
        "recall": recall_score(y_cv_val, y_cv_pred),
        "f1": f1_score(y_cv_val, y_cv_pred),
        "roc_auc": roc_auc_score(y_cv_val, y_cv_score),
        "pr_auc": average_precision_score(y_cv_val, y_cv_score)
    }

    cv_results_fixed.append(fold_result)

    print(f"Accuracy : {fold_result['accuracy']:.4f}")
    print(f"Precision: {fold_result['precision']:.4f}")
    print(f"Recall   : {fold_result['recall']:.4f}")
    print(f"F1       : {fold_result['f1']:.4f}")
    print(f"ROC-AUC  : {fold_result['roc_auc']:.4f}")
    print(f"PR-AUC   : {fold_result['pr_auc']:.4f}")


===== Fold 1 =====
Accuracy : 0.9906
Precision: 0.8828
Recall   : 0.9275
F1       : 0.9046
ROC-AUC  : 0.9958
PR-AUC   : 0.9643

===== Fold 2 =====
Accuracy : 0.9860
Precision: 0.8278
Recall   : 0.8993
F1       : 0.8621
ROC-AUC  : 0.9955
PR-AUC   : 0.9155

===== Fold 3 =====
Accuracy : 0.9857
Precision: 0.8403
Recall   : 0.8705
F1       : 0.8551
ROC-AUC  : 0.9906
PR-AUC   : 0.9360

===== Fold 4 =====
Accuracy : 0.9878
Precision: 0.8662
Recall   : 0.8849
F1       : 0.8754
ROC-AUC  : 0.9922
PR-AUC   : 0.9432

===== Fold 5 =====
Accuracy : 0.9839
Precision: 0.8194
Recall   : 0.8551
F1       : 0.8369
ROC-AUC  : 0.9917
PR-AUC   : 0.9289


In [78]:
import pandas as pd

cv_df = pd.DataFrame(cv_results_fixed)

print("===== 5-Fold CV Summary =====")

for metric in ["accuracy", "precision", "recall", "f1", "roc_auc", "pr_auc"]:
    mean = cv_df[metric].mean()
    std = cv_df[metric].std()

    print(f"{metric.upper():10s}: {mean:.4f} ± {std:.4f}")

===== 5-Fold CV Summary =====
ACCURACY  : 0.9868 ± 0.0025
PRECISION : 0.8473 ± 0.0265
RECALL    : 0.8875 ± 0.0278
F1        : 0.8668 ± 0.0253
ROC_AUC   : 0.9931 ± 0.0023
PR_AUC    : 0.9376 ± 0.0181


In [79]:
import numpy as np
import pandas as pd
from sklearn.model_selection import StratifiedKFold
from sklearn.svm import LinearSVC
from sklearn.metrics import (
    precision_score,
    recall_score,
    f1_score
)

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

threshold_results = []

thresholds = np.arange(-0.50, 0.51, 0.01)

for fold, (train_idx_cv, val_idx_cv) in enumerate(
    skf.split(X_train_char_combined, y_train), 1
):

    print(f"\n===== Fold {fold} =====")

    X_cv_train = X_train_char_combined[train_idx_cv]
    X_cv_val = X_train_char_combined[val_idx_cv]

    y_cv_train = y_train.iloc[train_idx_cv]
    y_cv_val = y_train.iloc[val_idx_cv]

    model_cv = LinearSVC(
        class_weight="balanced",
        C=1.0,
        max_iter=5000,
        random_state=42
    )

    model_cv.fit(X_cv_train, y_cv_train)

    y_cv_score = model_cv.decision_function(X_cv_val)

    best_f1 = -1
    best_threshold = None
    best_precision = None
    best_recall = None

    for threshold in thresholds:

        y_cv_pred = (y_cv_score >= threshold).astype(int)

        precision = precision_score(
            y_cv_val,
            y_cv_pred,
            zero_division=0
        )

        recall = recall_score(
            y_cv_val,
            y_cv_pred,
            zero_division=0
        )

        f1 = f1_score(
            y_cv_val,
            y_cv_pred,
            zero_division=0
        )

        if f1 > best_f1:
            best_f1 = f1
            best_threshold = threshold
            best_precision = precision
            best_recall = recall

    threshold_results.append({
        "fold": fold,
        "best_threshold": best_threshold,
        "precision": best_precision,
        "recall": best_recall,
        "f1": best_f1
    })

    print(f"Best Threshold: {best_threshold:.2f}")
    print(f"Precision     : {best_precision:.4f}")
    print(f"Recall        : {best_recall:.4f}")
    print(f"F1            : {best_f1:.4f}")


===== Fold 1 =====
Best Threshold: -0.09
Precision     : 0.9528
Recall        : 0.8768
F1            : 0.9132

===== Fold 2 =====
Best Threshold: -0.09
Precision     : 0.8963
Recall        : 0.8705
F1            : 0.8832

===== Fold 3 =====
Best Threshold: -0.10
Precision     : 0.9070
Recall        : 0.8417
F1            : 0.8731

===== Fold 4 =====
Best Threshold: -0.29
Precision     : 0.8681
Recall        : 0.8993
F1            : 0.8834

===== Fold 5 =====
Best Threshold: 0.05
Precision     : 0.9487
Recall        : 0.8043
F1            : 0.8706


In [80]:
threshold_df = pd.DataFrame(threshold_results)

print("===== Optimized Threshold CV Summary =====")

for metric in ["best_threshold", "precision", "recall", "f1"]:
    mean = threshold_df[metric].mean()
    std = threshold_df[metric].std()

    print(f"{metric.upper():15s}: {mean:.4f} ± {std:.4f}")

===== Optimized Threshold CV Summary =====
BEST_THRESHOLD : -0.1040 ± 0.1212
PRECISION      : 0.9146 ± 0.0360
RECALL         : 0.8585 ± 0.0366
F1             : 0.8847 ± 0.0170


In [81]:
from sklearn.model_selection import StratifiedKFold
from sklearn.svm import LinearSVC
import numpy as np

skf = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

oof_scores = np.zeros(len(y_train))

for fold, (train_idx_cv, val_idx_cv) in enumerate(
    skf.split(X_train_char_combined, y_train), 1
):

    print(f"\n===== OOF Fold {fold} =====")

    X_cv_train = X_train_char_combined[train_idx_cv]
    X_cv_val = X_train_char_combined[val_idx_cv]

    y_cv_train = y_train.iloc[train_idx_cv]

    model_oof = LinearSVC(
        class_weight="balanced",
        C=1.0,
        max_iter=5000,
        random_state=42
    )

    model_oof.fit(X_cv_train, y_cv_train)

    oof_scores[val_idx_cv] = model_oof.decision_function(X_cv_val)

print("\nOOF scores generated:", len(oof_scores))


===== OOF Fold 1 =====

===== OOF Fold 2 =====

===== OOF Fold 3 =====

===== OOF Fold 4 =====

===== OOF Fold 5 =====

OOF scores generated: 14304


In [82]:
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np

thresholds = np.arange(-0.50, 0.51, 0.01)

oof_threshold_results = []

for threshold in thresholds:

    oof_pred = (oof_scores >= threshold).astype(int)

    precision = precision_score(
        y_train,
        oof_pred,
        zero_division=0
    )

    recall = recall_score(
        y_train,
        oof_pred,
        zero_division=0
    )

    f1 = f1_score(
        y_train,
        oof_pred,
        zero_division=0
    )

    oof_threshold_results.append({
        "threshold": threshold,
        "precision": precision,
        "recall": recall,
        "f1": f1
    })

oof_threshold_df = pd.DataFrame(oof_threshold_results)

best_oof = oof_threshold_df.loc[
    oof_threshold_df["f1"].idxmax()
]

print("===== OOF Threshold Selection =====")
print(f"Best Threshold: {best_oof['threshold']:.2f}")
print(f"Precision     : {best_oof['precision']:.4f}")
print(f"Recall        : {best_oof['recall']:.4f}")
print(f"F1            : {best_oof['f1']:.4f}")

===== OOF Threshold Selection =====
Best Threshold: -0.09
Precision     : 0.9161
Recall        : 0.8514
F1            : 0.8826


In [83]:
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    accuracy_score,
    roc_auc_score,
    average_precision_score
)

# Train final model on ALL training data
final_model = LinearSVC(
    class_weight="balanced",
    C=1.0,
    max_iter=5000,
    random_state=42
)

final_model.fit(X_train_char_combined, y_train)

# Test scores
y_test_score_final = final_model.decision_function(
    X_test_char_combined
)

# LOCKED OOF threshold
final_threshold = -0.09

y_test_pred_final = (
    y_test_score_final >= final_threshold
).astype(int)

print("===== FINAL TEST RESULTS =====")

print(
    classification_report(
        y_test,
        y_test_pred_final,
        digits=4
    )
)

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_test_pred_final))

print(f"Accuracy : {accuracy_score(y_test, y_test_pred_final):.4f}")
print(f"ROC-AUC  : {roc_auc_score(y_test, y_test_score_final):.4f}")
print(f"PR-AUC   : {average_precision_score(y_test, y_test_score_final):.4f}")

===== FINAL TEST RESULTS =====
              precision    recall  f1-score   support

           0     0.9953    0.9985    0.9969      3403
           1     0.9691    0.9075    0.9373       173

    accuracy                         0.9941      3576
   macro avg     0.9822    0.9530    0.9671      3576
weighted avg     0.9940    0.9941    0.9940      3576

Confusion Matrix:
[[3398    5]
 [  16  157]]
Accuracy : 0.9941
ROC-AUC  : 0.9911
PR-AUC   : 0.9684


In [84]:
import pandas as pd
import numpy as np

# Create error analysis dataframe
error_df = pd.DataFrame({
    "y_true": np.array(y_test),
    "y_pred": y_test_pred_final,
    "score": y_test_score_final
})

# Identify errors
errors = error_df[
    error_df["y_true"] != error_df["y_pred"]
].copy()

# Error type
errors["error_type"] = np.where(
    (errors["y_true"] == 0) & (errors["y_pred"] == 1),
    "False Positive",
    "False Negative"
)

print("===== ERROR ANALYSIS =====")
print(f"Total errors: {len(errors)}")
print(f"False Positives: {(errors['error_type'] == 'False Positive').sum()}")
print(f"False Negatives: {(errors['error_type'] == 'False Negative').sum()}")

print("\nErrors:")
print(errors)

===== ERROR ANALYSIS =====
Total errors: 21
False Positives: 5
False Negatives: 16

Errors:
      y_true  y_pred     score      error_type
167        1       0 -0.721948  False Negative
727        1       0 -0.379401  False Negative
904        1       0 -0.159318  False Negative
1006       1       0 -0.782836  False Negative
1350       1       0 -0.417897  False Negative
1452       1       0 -1.127974  False Negative
1800       1       0 -0.165689  False Negative
2123       1       0 -0.325279  False Negative
2130       1       0 -0.300214  False Negative
2200       0       1 -0.053507  False Positive
2219       1       0 -0.133154  False Negative
2254       0       1  0.009576  False Positive
2268       1       0 -1.368226  False Negative
2460       1       0 -0.578325  False Negative
2680       1       0 -0.656953  False Negative
2763       0       1  0.040237  False Positive
2869       1       0 -1.488418  False Negative
3162       0       1  0.247653  False Positive
3163       1   

In [86]:
# Get the positions of the misclassified samples
error_positions = errors.index.to_numpy()

# Extract the corresponding test samples
X_test_errors = X_test.iloc[error_positions].copy()

# Add actual/predicted labels and scores
X_test_errors["y_true"] = np.asarray(y_test)[error_positions]
X_test_errors["y_pred"] = np.asarray(y_test_pred_final)[error_positions]
X_test_errors["score"] = np.asarray(y_test_score_final)[error_positions]

print("===== MISCLASSIFIED SAMPLES =====")
display(X_test_errors)

===== MISCLASSIFIED SAMPLES =====


17648     Vemma Brand Partner Unknown Looking for motiva...
17598     Real Estate & Insurance Professionals Elite Re...
16860     Process Validation Project Manager  Maynard Co...
17808     vacancies,nannies,cooks,drivers and housekeepe...
5926      Assistant Accountant Unknown Looking for an As...
3463      Business Development Manager Unknown We are cu...
6144      Physician Assistant PA Dermatology   1843 Unkn...
17554     Entry Level Medical Transcriptionist Unknown W...
17644     Fidelity Unknown The key to this position is t...
9326      Big Data - Hadoop Designer Unknown Position: B...
15238     Graphite Expert (Remote Position) Unknown Posi...
11853     Outside Sale Rep Unknown We are looking for cr...
5672      Software Design Engineer Unknown Software Desi...
4761      COMPANY REPRESENTATIVE NEEDED Unknown Our moth...
5499      Brand Partner Unknown Looking for motivated an...
11361     Field Representative/ Hiring Immediately Unkno...
17667     Quant Analyst  Unknown Our cli

In [87]:
# Extract the misclassified text samples
error_positions = errors.index.to_numpy()

error_analysis = pd.DataFrame({
    "text": X_test.iloc[error_positions].to_numpy(),
    "y_true": np.asarray(y_test)[error_positions],
    "y_pred": np.asarray(y_test_pred_final)[error_positions],
    "score": np.asarray(y_test_score_final)[error_positions]
})

error_analysis["error_type"] = np.where(
    (error_analysis["y_true"] == 0) & (error_analysis["y_pred"] == 1),
    "False Positive",
    "False Negative"
)

display(error_analysis)

,text,y_true,y_pred,score,error_type
0,Vemma Brand Partner Unknown Looking for motiva...,1,0,-0.721948,False Negative
1,Real Estate & Insurance Professionals Elite Re...,1,0,-0.379401,False Negative
2,Process Validation Project Manager Maynard Co...,1,0,-0.159318,False Negative
3,"vacancies,nannies,cooks,drivers and housekeepe...",1,0,-0.782836,False Negative
4,Assistant Accountant Unknown Looking for an As...,1,0,-0.417897,False Negative
5,Business Development Manager Unknown We are cu...,1,0,-1.127974,False Negative
6,Physician Assistant PA Dermatology 1843 Unkn...,1,0,-0.165689,False Negative
7,Entry Level Medical Transcriptionist Unknown W...,1,0,-0.325279,False Negative
8,Fidelity Unknown The key to this position is t...,1,0,-0.300214,False Negative
9,Big Data - Hadoop Designer Unknown Position: B...,0,1,-0.053507,False Positive


In [88]:
pd.set_option("display.max_colwidth", None)

display(error_analysis[[
    "text",
    "y_true",
    "y_pred",
    "score",
    "error_type"
]])

text  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

In [89]:
for i, row in error_analysis.iterrows():
    print("=" * 100)
    print(f"ERROR {i}")
    print(f"True: {row['y_true']} | Pred: {row['y_pred']} | Score: {row['score']:.4f}")
    print(f"Type: {row['error_type']}")
    print("\nTEXT:")
    print(row["text"])
    print()

ERROR 0
True: 1 | Pred: 0 | Score: -0.7219
Type: False Negative

TEXT:
Vemma Brand Partner Unknown Looking for motivated and hard working individuals!*Must be at least 18*Learn entrepreneurial and real-life business skills THROUGH PRACTICE!Study Success and Follow successful people!Vemma is a health and nutrition company on the rise!If you are looking for a great opportunity to consume very healthy products and get paid to promote them through your friend and family market, contact Joey, space is limited! Motivated and willing to work with a team! Unknown

ERROR 1
True: 1 | Pred: 0 | Score: -0.3794
Type: False Negative

TEXT:
Real Estate & Insurance Professionals Elite Realty Services was formed in 2004 by its Founder, Robert Do, who originally owned and operated several Coldwell Banker offices.  Realizing the Franchise had some limitations in their services to the public and to its agents, he decided to form Elite Realty Services. His vision for the new company was to offer more to me

In [90]:
print("Class distribution:")
print(y.value_counts())

print("\nClass percentages:")
print(y.value_counts(normalize=True) * 100)

Class distribution:
fraudulent
0    17014
1      866
Name: count, dtype: int64

Class percentages:
fraudulent
0    95.1566
1     4.8434
Name: proportion, dtype: float64


In [91]:
# Show a few examples from each class
for label in sorted(y.unique()):
    print("\n" + "=" * 80)
    print(f"CLASS {label}")
    print("=" * 80)

    samples = X.loc[y == label].head(5)

    for i, text in samples.items():
        print("\n---")
        print(text[:1000])


CLASS 0

---
Marketing Intern We're Food52, and we've created a groundbreaking and award-winning cooking site. We support, connect, and celebrate home cooks, and give them everything they need in one place.We have a top editorial, business, and engineering team. We're focused on using technology to find new and better ways to connect people around their specific food interests, and to offer them superb, highly curated information about food and cooking. We attract the most talented home cooks and contributors in the country; we also publish well-known professionals like Mario Batali, Gwyneth Paltrow, and Danny Meyer. And we have partnerships with Whole Foods Market and Random House.Food52 has been named the best food website by the James Beard Foundation and IACP, and has been featured in the New York Times, NPR, Pando Daily, TechCrunch, and on the Today Show.We're located in Chelsea, in New York City. Food52, a fast-growing, James Beard Award-winning online food community and crowd-s

In [92]:
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    average_precision_score
)

# OOF predictions using your locked threshold
oof_pred = (oof_scores >= -0.09).astype(int)

print("OOF Metrics")
print("=" * 50)

print("Accuracy :", accuracy_score(y_train, oof_pred))
print("Precision:", precision_score(y_train, oof_pred))
print("Recall   :", recall_score(y_train, oof_pred))
print("F1       :", f1_score(y_train, oof_pred))
print("ROC-AUC  :", roc_auc_score(y_train, oof_scores))
print("PR-AUC   :", average_precision_score(y_train, oof_scores))

OOF Metrics
Accuracy : 0.9890240492170023
Precision: 0.9161490683229814
Recall   : 0.8513708513708513
F1       : 0.8825729244577412
ROC-AUC  : 0.9931041048519559
PR-AUC   : 0.937840128100447


In [93]:
print("FINAL MODEL")
print("=" * 50)
print("Model:", final_model)
print("Threshold:", -0.09)
print()
print("Test Accuracy :", accuracy_score(y_test, y_test_pred_final))
print("Test ROC-AUC  :", roc_auc_score(y_test, y_test_score_final))
print("Test PR-AUC   :", average_precision_score(y_test, y_test_score_final))

FINAL MODEL
Model: LinearSVC(class_weight='balanced', max_iter=5000, random_state=42)
Threshold: -0.09

Test Accuracy : 0.9941275167785235
Test ROC-AUC  : 0.9911197022688243
Test PR-AUC   : 0.9684204192390946


In [94]:
import joblib

final_artifact = {
    "model": final_model,
    "threshold": -0.09,
    "model_name": "LinearSVC",
}

joblib.dump(final_artifact, "final_model.joblib")

print("Model saved successfully.")

Model saved successfully.


# Final Model Evaluation & Conclusion

## Final Model
- **Model:** LinearSVC
- **Class Weight:** Balanced
- **Max Iterations:** 5000
- **Random State:** 42
- **Decision Threshold:** -0.09

## Validation & Test Performance

| Metric | OOF Validation | Final Test |
|---|---:|---:|
| Accuracy | 98.90% | **99.41%** |
| ROC-AUC | 99.31% | **99.11%** |
| PR-AUC | 93.78% | **96.84%** |

## Final Test Confusion Matrix

```text
[[3398    5]
 [  16  157]]